pkg/controller/client.go

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright 2024 The Kmesh Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	service_discovery_v3 "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"google.golang.org/grpc"
+
+	"kmesh.net/kmesh/pkg/controller/config"
+	"kmesh.net/kmesh/pkg/controller/envoy"
+	"kmesh.net/kmesh/pkg/controller/workload"
+	"kmesh.net/kmesh/pkg/nets"
+)
+
+const (
+	RandTimeSed = 1000
+)
+
+type XdsClient struct {
+	ctx            context.Context
+	cancel         context.CancelFunc
+	grpcConn       *grpc.ClientConn
+	client         service_discovery_v3.AggregatedDiscoveryServiceClient
+	adsStream      *envoy.AdsStream
+	workloadStream *workload.WorkloadStream
+	xdsConfig      *config.XdsConfig
+}
+
+func NewXdsClient() *XdsClient {
+	client := &XdsClient{
+		xdsConfig: config.GetConfig(),
+		adsStream: &envoy.AdsStream{
+			Event: envoy.NewServiceEvent(),
+		},
+		workloadStream: &workload.WorkloadStream{
+			Event: workload.NewServiceEvent(),
+		},
+	}
+
+	client.ctx, client.cancel = context.WithCancel(context.Background())
+
+	return client
+}
+
+func (c *XdsClient) createStreamClient() error {
+	var err error
+
+	if c.grpcConn, err = nets.GrpcConnect(c.xdsConfig.DiscoveryAddress); err != nil {
+		return fmt.Errorf("grpc connect failed: %s", err)
+	}
+
+	c.client = service_discovery_v3.NewAggregatedDiscoveryServiceClient(c.grpcConn)
+
+	if bpfConfig.EnableKmesh {
+		if err = c.adsStream.AdsCreateStream(c.client, c.ctx); err != nil {
+			return fmt.Errorf("create ads stream failed, %s", err)
+		}
+	} else if bpfConfig.EnableKmeshWorkload {
+		if err = c.workloadStream.WorklaodCreateStream(c.client, c.ctx); err != nil {
+			return fmt.Errorf("create workload stream failed, %s", err)
+		}
+	}
+
+	return nil
+}
+
+func (c *XdsClient) recoverConnection() error {
+	var (
+		err      error
+		interval = time.Second
+	)
+
+	c.closeStreamClient()
```

Comment:
why need close here, IMO, we should call this function after stream/connection failed

Reply:
When stream failed, we should close the grpc connection immediately

Reply:
> why need close here, IMO, we should call this function after stream/connection failed

close the currently failed connection and then reconnect

Reply:
You should close the connection in clientResponseProcess, and after error occurs

Reply:
fixed